File: amazon_scraper_generic.js
const { chromium } = require('playwright');
const ObjectsToCsv = require('objects-to-csv');

async function scrapeAmazon(url, outputFilename = 'amazon_output.csv') {
  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    viewport: { width: 1920, height: 1080 },
    locale: 'en-GB',
    timezoneId: 'Europe/London'
  });
  
  const page = await context.newPage();
  
  try {
    console.log(`Navigating to ${url}...`);
    await page.goto(url, {
      waitUntil: 'networkidle',
      timeout: 60000
    });
    
    console.log('Waiting for page to load...');
    
    try {
      await page.click('#sp-cc-accept', { timeout: 3000 });
      console.log('Cookie banner dismissed');
      await page.waitForTimeout(2000);
    } catch (e) {
      console.log('No cookie banner found or already dismissed');
    }
    
    try {
      await page.waitForSelector('[class*="ProductGridItem"], [data-component-type="s-search-result"]', { timeout: 10000 });
      console.log('Product grid loaded');
    } catch (e) {
      console.log('Product grid not found, continuing anyway');
    }
    
    await page.waitForTimeout(3000);
    
    console.log('Extracting product data...');
    
    const scrapedData = await page.evaluate(() => {
      const products = [];
      
      const productElements = document.querySelectorAll('[class*="ProductGridItem"], [data-component-type="s-search-result"], [data-asin]:not([data-asin=""])');
      
      productElements.forEach((element) => {
        let titleElement = element.querySelector('h1, h2, h3, h4, h5, h6');
        if (!titleElement) {
          titleElement = element.querySelector('[class*="title"], [class*="Title"]');
        }
        if (!titleElement) {
          const link = element.querySelector('a[href*="/dp/"]');
          if (link) titleElement = link;
        }
        
        let priceElement = element.querySelector('[class*="price"], [class*="Price"]');
        if (!priceElement) {
          priceElement = element.querySelector('.a-price, .a-price-whole, .a-offscreen');
        }
        
        const imageElement = element.querySelector('img');
        const linkElement = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
        
        const title = titleElement ? titleElement.textContent.trim() : '';
        const price = priceElement ? priceElement.textContent.trim() : '';
        const imageUrl = imageElement ? (imageElement.src || imageElement.dataset.src || imageElement.getAttribute('data-src') || '') : '';
        const productLink = linkElement ? linkElement.href : '';
        
        let asin = element.getAttribute('data-asin') || '';
        if (!asin && linkElement && linkElement.href) {
          const asinMatch = linkElement.href.match(/\/dp\/([A-Z0-9]{10})/);
          if (asinMatch) {
            asin = asinMatch[1];
          }
        }
        
        if (title || price || imageUrl || productLink) {
          products.push({
            title: title,
            price: price,
            imageUrl: imageUrl,
            productLink: productLink,
            asin: asin
          });
        }
      });
      
      if (products.length === 0) {
        const allElements = document.querySelectorAll('div, article, section');
        let foundCount = 0;
        
        allElements.forEach((element) => {
          if (foundCount >= 50) return;
          
          const hasImage = element.querySelector('img');
          const hasLink = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
          const hasPrice = element.querySelector('[class*="price"], [class*="Price"]') || 
                          element.textContent.match(/£\d+|\$\d+/);
          
          if (hasImage && (hasLink || hasPrice)) {
            const titleElement = element.querySelector('h1, h2, h3, h4, h5, h6, [class*="title"], [class*="name"]');
            const priceElement = element.querySelector('[class*="price"], [class*="Price"]');
            const imageElement = element.querySelector('img');
            const linkElement = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
            
            const title = titleElement ? titleElement.textContent.trim() : '';
            const price = priceElement ? priceElement.textContent.trim() : '';
            const imageUrl = imageElement ? (imageElement.src || imageElement.dataset.src || '') : '';
            const productLink = linkElement ? linkElement.href : '';
            
            if (title && title.length > 3) {
              products.push({
                title: title,
                price: price,
                imageUrl: imageUrl,
                productLink: productLink,
                asin: ''
              });
              foundCount++;
            }
          }
        });
      }
      
      return products;
    });
    
    console.log(`Found ${scrapedData.length} products`);
    
    const uniqueProducts = [];
    const seen = new Set();
    
    for (const product of scrapedData) {
      const key = `${product.title}-${product.asin}-${product.productLink}`;
      if (!seen.has(key) && (product.title || product.asin)) {
        seen.add(key);
        uniqueProducts.push(product);
      }
    }
    
    console.log(`After removing duplicates: ${uniqueProducts.length} products`);
    
    if (uniqueProducts.length > 0) {
      const csv = new ObjectsToCsv(uniqueProducts);
      await csv.toDisk(`./${outputFilename}`);
      console.log(`Data saved to ${outputFilename}`);
    } else {
      console.log('No products found to save');
    }
    
    console.log('\nSample data:');
    console.log(JSON.stringify(uniqueProducts.slice(0, 3), null, 2));
    
    return uniqueProducts;
    
  } catch (error) {
    console.error('Error during scraping:', error);
    throw error;
  } finally {
    await browser.close();
  }
}

if (require.main === module) {
  const url = process.argv[2];
  const outputFile = process.argv[3] || 'amazon_output.csv';
  
  if (!url) {
    console.error('Usage: node amazon_scraper_generic.js <amazon_url> [output_filename.csv]');
    console.error('Example: node amazon_scraper_generic.js "https://www.amazon.co.uk/stores/page/..." my_output.csv');
    process.exit(1);
  }
  
  scrapeAmazon(url, outputFile)
    .then(() => {
      console.log('Scraping completed successfully!');
      process.exit(0);
    })
    .catch((error) => {
      console.error('Scraping failed:', error);
      process.exit(1);
    });
}

module.exports = { scrapeAmazon };

File: src/scrapers/amazon_scraper_generic.py
#!/usr/bin/env python3
from playwright.sync_api import sync_playwright
import csv
import json
import sys

def scrape_amazon(url, output_filename='amazon_output.csv'):
    """
    Scrape Amazon product listings from any Amazon URL
    
    Args:
        url: Amazon URL to scrape
        output_filename: Output CSV filename (default: amazon_output.csv)
    
    Returns:
        List of product dictionaries
    """
    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=True,
            args=['--no-sandbox', '--disable-setuid-sandbox']
        )
        
        context = browser.new_context(
            user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            viewport={'width': 1920, 'height': 1080},
            locale='en-GB',
            timezone_id='Europe/London'
        )
        
        page = context.new_page()
        
        try:
            print(f'Navigating to {url}...')
            page.goto(url, wait_until='networkidle', timeout=60000)
            
            print('Waiting for page to load...')
            
            try:
                page.click('#sp-cc-accept', timeout=3000)
                print('Cookie banner dismissed')
                page.wait_for_timeout(2000)
            except Exception as e:
                print('No cookie banner found or already dismissed')
            
            try:
                page.wait_for_selector('[class*="ProductGridItem"], [data-component-type="s-search-result"]', timeout=10000)
                print('Product grid loaded')
            except Exception as e:
                print('Product grid not found, continuing anyway')
            
            page.wait_for_timeout(3000)
            
            print('Extracting product data...')
            
            scraped_data = page.evaluate('''() => {
                const products = [];
                
                const productElements = document.querySelectorAll('[class*="ProductGridItem"], [data-component-type="s-search-result"], [data-asin]:not([data-asin=""])');
                
                productElements.forEach((element) => {
                    let titleElement = element.querySelector('h1, h2, h3, h4, h5, h6');
                    if (!titleElement) {
                        titleElement = element.querySelector('[class*="title"], [class*="Title"]');
                    }
                    if (!titleElement) {
                        const link = element.querySelector('a[href*="/dp/"]');
                        if (link) titleElement = link;
                    }
                    
                    let priceElement = element.querySelector('[class*="price"], [class*="Price"]');
                    if (!priceElement) {
                        priceElement = element.querySelector('.a-price, .a-price-whole, .a-offscreen');
                    }
                    
                    const imageElement = element.querySelector('img');
                    const linkElement = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
                    
                    const title = titleElement ? titleElement.textContent.trim() : '';
                    const price = priceElement ? priceElement.textContent.trim() : '';
                    const imageUrl = imageElement ? (imageElement.src || imageElement.dataset.src || imageElement.getAttribute('data-src') || '') : '';
                    const productLink = linkElement ? linkElement.href : '';
                    
                    let asin = element.getAttribute('data-asin') || '';
                    if (!asin && linkElement && linkElement.href) {
                        const asinMatch = linkElement.href.match(/\\/dp\\/([A-Z0-9]{10})/);
                        if (asinMatch) {
                            asin = asinMatch[1];
                        }
                    }
                    
                    if (title || price || imageUrl || productLink) {
                        products.push({
                            title: title,
                            price: price,
                            imageUrl: imageUrl,
                            productLink: productLink,
                            asin: asin
                        });
                    }
                });
                
                if (products.length === 0) {
                    const allElements = document.querySelectorAll('div, article, section');
                    let foundCount = 0;
                    
                    allElements.forEach((element) => {
                        if (foundCount >= 50) return;
                        
                        const hasImage = element.querySelector('img');
                        const hasLink = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
                        const hasPrice = element.querySelector('[class*="price"], [class*="Price"]') || 
                                        element.textContent.match(/£\\d+|\\$\\d+/);
                        
                        if (hasImage && (hasLink || hasPrice)) {
                            const titleElement = element.querySelector('h1, h2, h3, h4, h5, h6, [class*="title"], [class*="name"]');
                            const priceElement = element.querySelector('[class*="price"], [class*="Price"]');
                            const imageElement = element.querySelector('img');
                            const linkElement = element.querySelector('a[href*="/dp/"], a[href*="/gp/"]');
                            
                            const title = titleElement ? titleElement.textContent.trim() : '';
                            const price = priceElement ? priceElement.textContent.trim() : '';
                            const imageUrl = imageElement ? (imageElement.src || imageElement.dataset.src || '') : '';
                            const productLink = linkElement ? linkElement.href : '';
                            
                            if (title && title.length > 3) {
                                products.push({
                                    title: title,
                                    price: price,
                                    imageUrl: imageUrl,
                                    productLink: productLink,
                                    asin: ''
                                });
                                foundCount++;
                            }
                        }
                    });
                }
                
                return products;
            }''')
            
            print(f'Found {len(scraped_data)} products')
            
            unique_products = []
            seen = set()
            
            for product in scraped_data:
                key = f"{product['title']}-{product['asin']}-{product['productLink']}"
                if key not in seen and (product['title'] or product['asin']):
                    seen.add(key)
                    unique_products.append(product)
            
            print(f'After removing duplicates: {len(unique_products)} products')
            
            if len(unique_products) > 0:
                with open(output_filename, 'w', newline='', encoding='utf-8') as csvfile:
                    fieldnames = ['title', 'price', 'imageUrl', 'productLink', 'asin']
                    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
                    
                    writer.writeheader()
                    for product in unique_products:
                        writer.writerow(product)
                
                print(f'Data saved to {output_filename}')
            else:
                print('No products found to save')
            
            print('\nSample data:')
            print(json.dumps(unique_products[:3], indent=2))
            
            return unique_products
            
        except Exception as error:
            print(f'Error during scraping: {error}')
            raise
        finally:
            browser.close()

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: python amazon_scraper_generic.py <amazon_url> [output_filename.csv]')
        print('Example: python amazon_scraper_generic.py "https://www.amazon.co.uk/stores/page/..." my_output.csv')
        sys.exit(1)
    
    url = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else 'amazon_output.csv'
    
    try:
        scrape_amazon(url, output_file)
        print('Scraping completed successfully!')
    except Exception as e:
        print(f'Scraping failed: {e}')
        sys.exit(1)
